docs/configuration.md

```diff
@@ -0,0 +1,297 @@
+# Cardano SL configuration
+
+This document describes Cardano SL configuration. Configuration is
+stored in YAML format, conventional name for the file is
+`configuration.yaml`. This file contains multiple configurations, each
+one is associated with a key. Almost all executables accept two
+options to specify configuration: `--configuration-file` is used to
+specify path to configuration and `--configuration-key` specifies key
+in this configuration. An example of configuration can be found in
+file `lib/configuration.yaml`.
+
+## Genesis
+
+Configuration is used to specify genesis data used by the node. It is
+part of `core` configuration (it's accessible by key
+`<configuration-key>.core.genesis`). There are two ways to specify
+genesis data.
+
+* One way is to provide genesis data itself, stored in json
+  format. Genesis data explicitly states balances of all addresses,
+  all bootstrap stakeholders, heavyweight delegation certificates,
+  etc. It makes it clearer which genesis is used, but it doesn't
+  describe _how_ this genesis was constructed. Also it requires more
+  steps to create genesis data, than to create a spec. Current design
+  implies that this mechanism should be used only for mainnet where we
+  want to have full understanding of which addresses have which
+  stake. The format is demonstrated below. `hash` is the hash (Blake2b
+  256 bits) of canonically encoded `mainnet-genesis.json` file. It can
+  be computed manually using `scripts/js/genesis-hash.js` script.
+
+```
+    genesis:
+      src:
+        file: mainnet-genesis.json
+        hash: 5f20df933584822601f9e3f8c024eb5eb252fe8cefb24d1317dc3d432e940ebb
+```
+
+* Another way is to provide a specification how to generate genesis
+  data. It allows one to specify how many nodes should have stake,
+  which bootstrap stakeholders should be used, what should be total
+  balance, etc. When node is launched with spec as genesis, it
+  automatically constructs genesis data from it. This way to specify
+  genesis is supposed to be used for everything except
+  mainnet. Example:
```

Comment:
..and also anything production-y, like `staging` and `testnet`, right?

Can't see how we could want to use auto-genesis for those, since we want to maximise similarity to mainnet.

Reply:
For staging – yes.
For testnet – we didn't plan to do it, because we need significantly different genesis for testnet anyway, so this similarity shouldn't really matter. However, at least for the first testnet we will be using the same way to specify genesis as for mainnet for performance reasons.
It's not yet described in the document, I am waiting for @georgeee to finalize the requirements (I was told about them only on Monday in the evening and only briefly).

Reply:
> we need significantly different genesis for testnet anyway

I'm interested in hearing more about this when the info is available. I share Serge's intuition that testnet, for the most part, should be as similar to mainnet's configuration as possible.

However, as long as the initial genesis data resulting from the spec is sufficiently similar to mainnet's JSON in terms of constants and such, I'm not terribly concerned about automating it. We would just need a way to verify what initial genesis data are used.

Reply:
cc @domenkozar, based on DevOps review.